### PR TITLE
feat: show proof hashes in each batch

### LINKED ISSUE #605: feat(explorer): show proof hashes in each batch
For this, the hash of each proof must be submitted to S3, rather than recalculated in the explorer side.
Then, the explorer must fetch, store, and show this data.

### PR DESCRIPTION
**Pull request by uri-99**
## Notes

* ~~Merge after #580~~

## Changes

* add new table ecto migrate for proof data
* calculate keccak of proof in explorer to get proof data

### Before

image

### After

https://github.com/user-attachments/assets/0e6dbedb-d1a1-48a2-b4b2-88101affed50

### CODE REVIEW

File: explorer/lib/explorer/models/batches.ex [hunk 1]
```diff
@@ -14,15 +14,16 @@ defmodule Batches do
     field :response_transaction_hash, :string
     field :response_timestamp, :utc_datetime
     field :data_pointer, :string
+    field :proof_hashes, {:array, :string}
```
Comment (JuArce) [inline explorer/lib/explorer/models/batches.ex]: Maybe `proof_hashes` should be stored using an 1-N relationship table.
It brings several improvements
- Hashes are indexables, so you can search for a proof and get the associated batch
- If you need to add more info to the proofs, a second table would be easier to use

File: Makefile [hunk 2]
```diff
@@ -217,7 +217,7 @@ batcher_send_sp1_burst:
 		--proving_system SP1 \
 		--proof ../../scripts/test_files/sp1/sp1_fibonacci.proof \
 		--vm_program ../../scripts/test_files/sp1/sp1_fibonacci.elf \
-		--repetitions 15 \
+		--repetitions 3 \
```
Comment (JuArce) [inline Makefile]: We shouldn't be changing these values on every pr

**Review by JuArce — CHANGES_REQUESTED** [review]

**Review by glpecile — APPROVED** [review]

File: explorer/lib/explorer_web/live/pages/batch/index.html.heex [hunk 3]
```diff
@@ -71,6 +71,25 @@
           text={"Just%20submitted%20a%20proof%20via%20@alignedlayer%0AI%20am%20now%20%23aligned%20✅%0A#{@site_url}/batches/#{@merkle_root}"}
         />
       <% end %>
+      Proofs in this batch:
+      <%= if @proof_hashes != :empty do %>
+        <%= if @proof_hashes == :nil do %>
+          <.button class="w-fit text-foreground" phx-click="hide_proofs">
+            <.icon name="hero-eye-slash" class="size-4" /> Explorer couldn't fetch proofs for this batch
+          </.button>
```
Comment (glpecile) [inline explorer/lib/explorer_web/live/pages/batch/index.html.heex]:
```suggestion
          <p>
            There are no proofs available for this batch.
          </p>
```

File: explorer/lib/explorer/models/proofs.ex [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  # def get_proof(%{proof_hash: proof_hash}) do
+  #   query = from(p in Proofs,
+  #   where: p.proof_hash == ^proof_hash,
+  #   select: p)
+
+  #   case Explorer.Repo.all(query) do
+  #     nil -> nil
+  #     result -> [result, result |> Enum.count()]
+  #   end
+  # end
```
Comment (glpecile) [inline explorer/lib/explorer/models/proofs.ex]: This code should probably be deleted

File: explorer/lib/explorer/periodically.ex [hunk 5]
```diff
@@ -60,10 +60,12 @@ defmodule Explorer.Periodically do
...
-        batch
-          |> Utils.extract_amount_of_proofs
-          |> Batches.generate_changeset
-          |> Batches.insert_or_update
+        {batch_changeset, proofs} =
+          batch
+            |> Utils.extract_info_from_data_pointer
+            |> Batches.generate_changesets
+
+        Batches.insert_or_update(batch_changeset, proofs) #handles it atomically
```
Comment (glpecile) [inline explorer/lib/explorer/periodically.ex:68]:
```suggestion
        batch_changeset |> Batches.insert_or_update(proofs)
```

**Review by glpecile — CHANGES_REQUESTED** [review]

**Review by glpecile — APPROVED** [review]

File: explorer/lib/explorer/aligned_layer_service_manager.ex [hunk 6]
```diff
@@ -133,6 +134,7 @@ defmodule AlignedLayerServiceManager do
           response_transaction_hash: batch_response.transaction_hash,
           response_timestamp: batch_response.block_timestamp,
           amount_of_proofs: unverified_batch.amount_of_proofs,
+          proof_hashes: nil #don't need this value to update an existing but unverified batch, it is on another table
```
Comment (JuArce) [inline explorer/lib/explorer/aligned_layer_service_manager.ex:137]: What does it mean?
> don't need this value to update an existing but unverified batch, it is on another table

File: explorer/lib/explorer_web/live/utils.ex [hunk 7]
```diff
@@ -123,12 +123,21 @@ defmodule Utils do
...
 
-  def extract_amount_of_proofs_from_json({:ok, batch_json}) do
-    batch_json |> Enum.count()
+  def calculate_proof_hashes({:ok, batch_json}) do
+    IO.inspect("Calculating proof hashes")
+    batch_json
+      |> Enum.map(
+        fn proof ->
+          :crypto.hash(:sha3_256, proof["proof"])
+            # TODO removed this because i want to store as bytea, not a string.
```
Comment (JuArce) [inline explorer/lib/explorer_web/live/utils.ex]: Is this TODO solved?

Comment (glpecile) [reply explorer/lib/explorer_web/live/utils.ex]: This is already done in the frontend to show the proofs so yes, it is resolved. I'll delete the commented code

File: explorer/lib/explorer_web/live/utils.ex [hunk 8]
```diff
@@ -123,12 +123,21 @@ defmodule Utils do
...
+            # |> Base.encode16(case: :lower)
+            # |> (&("0x" <> &1)).()
```
Comment (glpecile) [inline explorer/lib/explorer_web/live/utils.ex]:
```suggestion
```

File: explorer/lib/explorer/aligned_layer_service_manager.ex (hunk 6 above)
Comment (uri-99) [reply explorer/lib/explorer/aligned_layer_service_manager.ex:137]: it means it's ok for that function to return the struct with `nil` in its proof_hashes parameter

**Review by NicolasRampoldi — APPROVED** [review]

**Review by JuArce — APPROVED** [review]

**Pull request merged**